internal/target/registry.go

```diff
@@ -0,0 +1,62 @@
+package target
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/boundary/internal/oplog"
+	"github.com/hashicorp/boundary/internal/types/subtypes"
+)
+
+type RepositoryHooks interface {
+	Alloc(publicId string, version uint32, op oplog.OpType) (Target, oplog.Metadata)
+	ValidateCreate(context.Context, Target) error
+	ValidateUpdate(context.Context, Target) error
+	NewTargetId() (string, error)
+}
+
+type repositoryHooksRegistry struct {
+	sync.Map
+}
+
+func (r *repositoryHooksRegistry) Set(s subtypes.Subtype, h RepositoryHooks) {
+	_, previouslySet := r.LoadOrStore(s, h)
+	if previouslySet {
+		panic(fmt.Sprintf("target subtype %s already registered", s))
+	}
+}
+
+func (r *repositoryHooksRegistry) Get(s subtypes.Subtype) (RepositoryHooks, bool) {
+	rh, ok := r.Load(s)
+	if ok {
+		return rh.(RepositoryHooks), ok
+	}
+	return nil, ok
+}
+
+var rhRegistry = repositoryHooksRegistry{sync.Map{}}
+var subtypeRegistry = subtypes.NewRegistry()
+
+// SubtypeFromType returns the Subtype from the provided string or if
+// no Subtype was registered with that string Unknown is returned.
+func SubtypeFromType(t string) subtypes.Subtype {
+	return subtypeRegistry.SubtypeFromType(t)
+}
+
+// SubtypeFromId returns the Subtype from the provided id if the id's prefix
+// was registered with a Subtype. Otherwise Unknown is returned.
+func SubtypeFromId(id string) subtypes.Subtype {
+	return subtypeRegistry.SubtypeFromId(id)
+}
+
+// Register registers repository hooks all the prefixes for a provided Subtype. Register returns
```

Comment:
Something is off with this wording, perhaps:
```suggestion
// Register registers repository hooks and all the prefixes for a provided Subtype. Register returns
```